### PR TITLE
Add an option to enable/disable user registration.

### PR DESCRIPTION
**Pull request by awyl**
#1252 

Mostly copy from O1.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using OrchardCore.Settings;
+
+namespace OrchardCore.Users.Models
+{
+    public class RegistrationSettings
+    {
+        public bool UsersCanRegister { get; set; }
+    }
+
+    public class RegistrationSettingsConfiguration : IConfigureOptions<RegistrationSettings>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs]: You don't need an `IOptions` as everything is in the site settings. Just get the site settings where you need it (in the controller).

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 2]
```diff
@@ -78,17 +80,27 @@ public class AccountController : Controller
 
         [HttpGet]
         [AllowAnonymous]
-        public IActionResult Register(string returnUrl = null)
+        public IActionResult Register([FromServices]IOptions<RegistrationSettings> regSettings, string returnUrl = null)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs]: I prefer the service to be injected in the ctor.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 3]
```diff
@@ -78,17 +80,27 @@ public class AccountController : Controller
...
         {
+            if ( !regSettings.Value.UsersCanRegister )
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs]: No spaces in the parenthesis.

File: src/OrchardCore.Modules/OrchardCore.Users/Drivers/RegistrationSettingsDisplayDriver.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace OrchardCore.Users.Drivers
+{
+    public class RegistrationSettingsDisplayDriver : SectionDisplayDriver<ISite, RegistrationSettings>
+    {
+        public const string GroupId = "registrationSettings";
+
+        public override IDisplayResult Edit(RegistrationSettings section)
+        {
+            return Shape<RegistrationSettings>( "RegistrationSettings_Edit", model => {
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Drivers/RegistrationSettingsDisplayDriver.cs]: You must have a global setting that adds spaces inside all parenthesizes ;)

**Comment by sebastienros** [comment]
Very nice effort. Have you looked at the template from ASP.NET core actually instead of O1 ? This is what we took for a start.

File: src/OrchardCore.Modules/OrchardCore.Users/Drivers/RegistrationSettingsDisplayDriver.cs (hunk 4 above)
Comment (awyl) [reply src/OrchardCore.Modules/OrchardCore.Users/Drivers/RegistrationSettingsDisplayDriver.cs]: Yes, easier to read for me. :)

File: src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs (hunk 1 above)
Comment (awyl) [reply src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs]: Done.

**Comment by awyl** [comment]
I haven't looked at it since 1.0.  Is there anything I should be aware of?

**Pull request merged**